tidy3d/components/geometry.py

```diff
@@ -3280,8 +3289,374 @@ def _surface_area(self, bounds: Bound) -> float:
+        solid_index : int
+            If set, read a single solid with this index from the file.
+
+        Returns
+        -------
+        Union[:class:`.CustomSurfaceMeshGeometry`, List[:class:`.GeometryGroup`]]
+            The geometry or geometries from the file.
+        """
+
+        def process_single(mesh: trimesh.Trimesh) -> CustomSurfaceMeshGeometry:
+            """Process a single trimesh using scale and origin."""
+            mesh.apply_scale(scale)
+            mesh.apply_translation(origin)
+            return cls.from_trimesh(mesh)
+
+        scene = trimesh.load(filename, **kwargs)
+        meshes = []
+        if isinstance(scene, trimesh.Trimesh):
+            meshes = [scene]
+        elif isinstance(scene.trimesh.Scene):
+            meshes = scene.dump()
+        else:
+            raise ValidationError(
+                "Invalid Trimesh type in file. Supported types are Trimesh and Scene."
+            )
+
+        if solid_index is None:
+            if isinstance(scene, trimesh.Trimesh):
+                return process_single(scene)
+            if isinstance(scene, trimesh.Scene):
+                geoms = [process_single(mesh) for mesh in meshes]
+                return GeometryGroup(geometries=geoms)
+
+        if solid_index < len(meshes):
+            return process_single(meshes[solid_index])
+        raise ValidationError("No solid found at 'solid_index' in the stl file.")
+
+    @classmethod
+    def from_trimesh(cls, mesh: trimesh.Trimesh) -> CustomSurfaceMeshGeometry:
+        """Create a :class:`.CustomSurfaceMeshGeometry` from a ``Trimesh`` object.
+
+        Parameters
+        ----------
+        trimesh : ``trimesh.Trimesh``
+            The Trimesh object containing the surface geometry mesh data.
+
+        Returns
+        -------
+        :class:`.CustomSurfaceMeshGeometry`
+            The custom surface mesh geometry given by the Trimesh provided.
+        """
+        return cls.from_vertices_faces(mesh.vertices, mesh.faces)
+
+    @classmethod
+    def from_triangles(cls, triangles: np.ndarray) -> CustomSurfaceMeshGeometry:
+        """Create a :class:`.CustomSurfaceMeshGeometry` from a numpy array
+        containing the triangles of a surface mesh.
+
+        Parameters
+        ----------
+        triangles : ``np.ndarray``
+            A numpy array of shape (N, 3, 3) storing the triangles of the surface mesh.
+            The first index labels the triangle, the second index labels the vertex
+            within a given triangle, and the third index is the coordinate (x, y, or z).
+
+        Returns
+        -------
+        :class:`.CustomSurfaceMeshGeometry`
+            The custom surface mesh geometry given by the triangles provided.
+
+        """
+        if len(triangles.shape) != 3 or triangles.shape[1] != 3 or triangles.shape[2] != 3:
+            raise ValidationError("Provided 'triangles' must be an N x 3 x 3 array.")
```

Comment:
```
f"Provided 'triangles' must be an N x 3 x 3 array, given {triangles.shape}."
```

Reply:
for some reason this isn't working, it's printing {triangles.shape} literally:

```vertices=[[0,0],[0,1],[1,0],[1,1]]
faces=[[1,2,3],[0,3,2],[0,1,3],[0,2,1]]
geom = td.CustomSurfaceMeshGeometry.from_vertices_faces(vertices, faces)```

```           ERROR    Provided 'vertices' must be an N x 3 array, given {vertices.shape}. ```